src/Control/Monad/Compat.hs

```diff
@@ -3,18 +3,49 @@ module Control.Monad.Compat (
   module Base
 , Monad(..)
 , MonadPlus(..)
-, void
+#if !MIN_VERSION_base(4,8,0)
+, foldM
+, foldM_
+, forM
+, forM_
+, guard
+, mapM
+, mapM_
+, msum
+, sequence
+, sequence_
+, unless
+, when
+
 , (<$!>)
-) where
-import Control.Monad as Base
+#endif
 
-#if !MIN_VERSION_base(4,3,0)
-import Data.Function (const)
-import Data.Functor (Functor(..))
+#if !MIN_VERSION_base(4,5,0)
+, void
 #endif
+) where
 
-#if !MIN_VERSION_base(4,8,0)
-import Prelude.Compat (seq)
+#if MIN_VERSION_base(4,8,0)
+import Control.Monad as Base
+#else
+import Control.Monad as Base hiding (
+    foldM
+  , foldM_
+  , forM
+  , forM_
+  , guard
+  , mapM
+  , mapM_
+  , msum
+  , sequence
+  , sequence_
+  , unless
+  , when
+  )
+import Control.Applicative.Compat (Alternative(..))
+import Data.Foldable.Compat
+import Data.Traversable.Compat
+import Prelude.Compat hiding (mapM, mapM_, sequence, sequence_)
```

Comment:
Is it necessary to hide `mapM`, etc.?  I thought prelude uses the generalized versions too?